Add v2 config validations to the build tool

jholecek-rh opened the pull request:
Regex validations are still missing. I will ad them in a separate PR after this one. This PR is already big enough.

See individual commits for details.

jholecek-rh committed:
Add schema validation for v2 remote configs to build.py

jholecek-rh committed:
Make build tool test case 'invalid_corrupted_v1_config_schema' unique

Using a unique string in the test will enable easier troubleshooting:
* Relevant files can be found using the unique string
* The string reinforces the test purpose

jholecek-rh committed:
Make build tool test case 'invalid_v1_config_fails_schema_validation' unique

Using a unique string in the test will enable easier troubleshooting:
* Relevant files can be found using the unique string
* The string reinforces the test purpose

jholecek-rh committed:
Improve build tool schema validation errors

Error messages should be actionable on their own. For the build tool,
this means they should state where the validated data came from and what
schema was used to validate it.

jholecek-rh committed:
Validate cluster version mapping schema in build tool

jholecek-rh committed:
Validate that cluster version mapping is sorted in build tool

Pairs in the cluster_version_mapping.json file must have strictly
increasing semantic versions.

jholecek-rh committed:
Validate first version in cluster version mapping in build tool

The cluster_version_mapping.json file has to cover all OCP version with
the Rapid Recommendations feature. That means that the first interval
must start at 4.17.0-0 or smaller version.

jholecek-rh committed:
Validate config references in cluster version mapping in build tool

jholecek-rh committed:
Refactor how build tool writes remote config files

The loop that writes configs for a given version API is a convenient
place to validate the generated configs. Since the validations differ
between v1 and v2 config files, we need a way to specify the validations
separately of each other.

ikerreyes commented on build.py:
```diff
@@ -127,15 +131,36 @@ def _write_v2(self, outputdir):
+
+    def _assert_cluster_version_mapping_order(self, filepath, content):
+        i = 0
+        v1 = semver.Version.parse(content[0][0])
```
Maybe we can start with v1 being 0.0 and then make the loop through the content, rather than starting at index 1? I find it more readable, but it is very subjective, so whatever you prefer.

jholecek-rh replied:
I didn't want to initialize the variable with any specific version because it would put additional constraints on the file. `0.0.0-0` is a valid version for the first item in the file.

I somehow thought that starting with `v1 = None` would make the code more convoluted, but it actually worked out quite nice. I've pushed the change.

ikerreyes replied:
OK, thank you.

ikerreyes commented on build.py:
```diff
@@ -127,15 +131,36 @@ def _write_v2(self, outputdir):
+    def _assert_cluster_version_mapping_order(self, filepath, content):
+        i = 0
+        v1 = semver.Version.parse(content[0][0])
+        for raw_version, _ in content[1:]:
```
Should we also assert that the second item in the tuple is bigger than the first one?

jholecek-rh replied:
No, for two reasons:

1. The first items in the pairs have to be strictly increasing, so the second items cannot influence the ordering.
2. The second items in the pairs are paths; file names should be descriptive of what the configuration does rather than OCP version ranges the files are used for.

ikerreyes replied:
Yeah, sorry. I forgot that we had ranges defined only by the first element in each line. My bad. Sorry, I was confused.

ikerreyes commented on build.py:
```diff
@@ -127,15 +131,36 @@ def _write_v2(self, outputdir):
+    def _assert_cluster_version_mapping_order(self, filepath, content):
+        i = 0
+        v1 = semver.Version.parse(content[0][0])
+        for raw_version, _ in content[1:]:
```
You can also use `enumerate` to get the index rather than having `i` as a separate variable.

jholecek-rh replied:
I didn't like enumerate when not starting the loop at index 0. I'm using it now since I refactored the code as discussed in the other thread. Please check again.

jholecek-rh committed:
Refactor cluster version mapping checks in build tool

The refactoring affects how the buid tool reports errors when items in
cluster_version_mapping.json are not sorted properly. Originally, it
reported the index of the last valid item. It reports the index of the
offending item now.

ikerreyes commented on build.py:
```diff
@@ -109,51 +113,108 @@ def write(self, outputdir):
+
+    def _assert_cluster_version_mapping_order(self, filepath, content):
+        v1 = None
+        for i, entry in enumerate(content):
```
I think you can also do `for i, (raw_version, _) in enumerate(content)`

jholecek-rh replied:
Indeed. Thanks! Done.

jholecek-rh committed:
Clean up build tool code
